Barbara solves computational problems with async (#132)
rust-lang/wg-async-foundations: src/vision/status_quo/barbara_simulates_hydrodynamics.md

> # 😱 Status quo stories: Barbara Builds a Hydrodynamics Simulator

Reviewer: ```suggestion
# 😱 Status quo stories: Niklaus Builds a Hydrodynamics Simulator
```

Reviewer: Have to rename the file, too

Author: Fixed

This new solution works, but Niklaus is not satisfied with how complex his code became after the move to `async` and that compilation time is now 30-60 seconds.  The state sharing adding a large amount of cruft with `Arc` and `async` is not well suited for using a dependency graph to schedule tasks so implementing this solution created a key component of his program that was difficult to understand and pervasive. Ultimately, his conclusion was that `async` is not appropriate for parallelizing computational tasks. He will be trying a new design based upon Rayon in the next version of her application.

## 🤔 Frequently Asked Questions

### **What are the morals of the story?**
- `async` looks to be the wrong choice for parallelizing compute bound/computational work
- There is a lack of guidance to help people solving such problems get started on the right foot
- Quality of Life issues (compilation time, type inference on `Result`) can create a drag on users ability to focus on their domain problem

### **What are the sources for this story?**
This story is based on the experience of building the [kilonova](https://github.com/clemson-cal/app-kilonova) hydrodynamics simulation solver.

> ### **Why did you choose Barbara and Grace to tell this story?**

Reviewer: ```suggestion
### **Why did you choose Niklaus and Grace to tell this story?**
```

Author: Fixed

I chose Niklaus as the primary character in this story because this work was driven by someone who only uses programming for a small part of their work. Grace was chosen as a supporting character because of that persons experience with C/C++ programming and to avoid repeating characters.

### **How would this story have played out differently for the other characters?**
- Alan: there's a good chance he would have already had experience working with either async workflows in another language or doing parallelization of compute bound tasks; and so would already know from experience that `async` was not the right place to start.  
- Grace: likewise, might already have experience with problems like this and would know what to look for when searching for tools. 

> - Niklaus: the experience would probably be the same, as it's very easy to assume that `tokio` is the starting place for concurrency in Rust.

Reviewer: ```suggestion
- Barbara: the experience would likely be fairly similar, since the actual subject of this story is quite familiar with Rust by now
```

Author: Fixed

Her first attempt to was to emulate a common CFD design pattern: using message passing to communicate between processes that are each assigned a specific patch in the grid. So she assign one thread to each patch and used messages to communicate solution state to dependent patches.  With one thread per patch this usually meant that there were 5-10x more threads than CPU cores.

This solution was fine, but Barbara was not satisified.  She had two problems with it: first, she didn't like that the design would greedily use all the resources on the machine and, second, when her team added a new feature (tracer particles) that increased the complexity of how patches interact with each other and the number of messages that were passsed between threads increased to the point that it became a performance bottleneck and parallel processing became no faster than serial processing. So, Barbara decided to find a better solution.

### Solution Path

> What Barbara wanted to do was find a way to more efficiently use threads: have a fixed number of threads that each mapped to a core on the CPU and assign patches to those threads as patches became ready to compute. The design of the `async` framework seemed to provide exactly that behavior. And to move away from the message passing design, because the number of messages being passed was proportional to the number of trace particles being traced.

Reviewer 2: What do you mean with the "design of the `async` framework"? Just async / await the language feature?

Author: Correct, I'll change this  to "The `async` feature seemed..."

Author: Fixed.

> As Barbara began working on her new design with `tokio`, her use of `async` went from a general (from the textbook) use of basic `async` features to a more specific implementation leveraging exactly the features that were most suited for her needs.

Reviewer 3: I'm surprised by Barbara's choice of `tokio` since `tokio` is normally meant for I/O bound workloads. Can you talk about the decision to use `tokio` more? Was it because Barbara needed an async runtime and `tokio` seemed like the only choice?

Reviewer: I think that it was just the most famous executor. @jzrake could confirm.

Reviewer 4: Yes, I (Barbara) was looking for a multi-threaded executor, and Tokio seemed to be the most mature. Along the way I learned that the async feature was never the right solution for data parallelism / CPU-bound tasks. With generous help from @erichgess I'm now working on a more appropriate solution.

I must say I don't believe Barbara is the right character for this story, because she is an experienced systems programmer. I (jzrake) am an astrophysicist, and have been learning Rust over the past <year from "the book", stack overflow, browsing GitHub, etc. It's only since contacting @nikomatsakis and now talking with @erichgess that I am getting guidance on industry practices. This little bit of interaction has already proven to be really productive.

There's a bigger moral here, which I hope is not being missed. There is no "Marvin the scientist" on the cast of characters, and I think that's symbolic of a neglected user base: the HPC community and scientists who badly need to modernize their codes. Many of us would love to adopt Rust, but there's very little intellectual capital at the intersection of Rust and scientific computing that could enable researchers to learn by osmosis. I see a lot of potential in the language, and my group here at Clemson is accumulating experience, which will be share with the HPC / physics simulation community. However widespread adoption of Rust in HPC is not going to happen until someone (Amazon maybe?) lends an ear to that community's needs, and invests in the types of training and educational materials that were made available to Python users by outfits like Enthought and Anaconda -- efforts which fueled the widespread adoption of Python (over Fortran and IDL) in observational astronomy and experimental physics.

Author: One quick note: Barbara is the experienced Rust developer (Grace is the experienced systems developer).  I went with Barbara here because from what I could tell you have some level of experience with Rust and comfortable with both Rust and learning more advanced concepts of programming.  However, I don't know if "experienced" for Barbara means deep knowledge of Rust or if it means that you've used Rust enough that you can understand lifetimes and borrow errors without having to use Google :smile:.

Barbara's priorities are:
```
    Top priority: overall productivity and long-term maintenance -- she loves Rust, and wants to see it extended to new areas; she has an existing code base to maintain
    Expectations: elegance and craftsmanship, fits well with Rust
```

Niklaus is an alternative that lines up with your not coming from a programming background, but I don't think you're close to being a beginner programmer.

Reviewer 4: Yes, thanks, I guess Barbara is the best fit of the available characters (I still kind of like Marvin but oh well). To me experienced means "productive" -- can get things done with the language. As opposed to "deep knowledge", which I'd think means you work on the compiler. Or at least do lots of good quality `unsafe`.

Reviewer: There may well be a missing character. Interesting. Marvin is kind of like Niklaus, but I do agree @jzrake that you seemed like someone who knew Rust fairly well.

Author: One thing about the 3 experienced characters is they differ in what types of software they write rather than how programming fits into their life.  When reading their descriptions, to me, it's pretty clear that they're all people who's job is writing software. But that doesn't cover people who only use programming as one of many tools to do their job.  @jzrake would fall into this group, along with data scientists, economists, and so on.

Reviewer: Having given this some thought, I think we ought to use Niklaus for this. I think we can generalize him to "non-programmers (yet)" of various backgrounds.

Author: Done.

> At first, Barbara was under a false impression about what async executors do. She had assumed that a multi-threaded executor could automatically move the execution of an async block to a worker thread. Then she discovered that async tasks must be explicitly spawned into into a thread pool if they are to be executed on a worker thread. This meant that the algorithm to be parallelized became strongly coupled to both the spawner and the executor. Code that used to cleanly express a physics algorithm now had interspersed references to the task spawner, not only making it harder to understand, but also making it impossible to try different execution strategies, since with Tokio the spawner and executor are the same object (the Tokio runtime). Barbara feels that a better design for data parallelism would enable better separation of concerns: a group of interdependent compute tasks, and a strategy to execute them in parallel.

Reviewer 2: Typo: "spawned into into a thread pool"

Author: Fixed.

> At first, Barbara was under a false impression about what async executors do. She had assumed that a multi-threaded executor could automatically move the execution of an async block to a worker thread. Then she discovered that async tasks must be explicitly spawned into into a thread pool if they are to be executed on a worker thread. This meant that the algorithm to be parallelized became strongly coupled to both the spawner and the executor. Code that used to cleanly express a physics algorithm now had interspersed references to the task spawner, not only making it harder to understand, but also making it impossible to try different execution strategies, since with Tokio the spawner and executor are the same object (the Tokio runtime). Barbara feels that a better design for data parallelism would enable better separation of concerns: a group of interdependent compute tasks, and a strategy to execute them in parallel.

Reviewer 2: How did she discover that an async task must be explicitly spawned into a threadpool?

Author: I'll follow up with the source to get details on how they figured this out.

In order to remove the need for message passing, Barbara moved to a shared state design: she would keep a table tracking the solution state for every grid patch and a specific patch would only start its computation task when solutions were written for all the patches it was dependent on. So, each task needs to access the table with the solution results of all the other tasks. Learning how to properly use shared data with `async` was a new challenge. The initial design:

```rust
    let mut stage_primitive_and_scalar = |index: BlockIndex, state: BlockState<C>, hydro: H, geometry: GridGeometry| {
        let stage = async move {
            let p = state.try_to_primitive(&hydro, &geometry)?;
            let s = state.scalar_mass / &geometry.cell_volumes / p.map(P::lorentz_factor);
            Ok::<_, HydroError>( ( p.to_shared(), s.to_shared() ) )
        };
        stage_map.insert(index, runtime.spawn(stage).map(|f| f.unwrap()).shared());
    };
```

> lacked performance because she needed to clone the value for every task.  So, Barbara switched over to using `Arc` to keep a thread safe RC to the shared data. But this change introduced a lot of `.map` and `.unwrap` function calls, making the code much harder to read. She realized that managing the dependency graph was not intuitive when using `async` for concurrency.

Reviewer 3: "she needed to clone _the value_ for every task" - which value are you referring to?

I'm having a hard to relating the code example and the problems mentioned in this paragraph. Perhaps comments in the code examples could bring better attention to the relevant details?

Reviewer 4: Yes, `shared` comes from the `futures` crate. I think the basic pattern I was trying can be represented more cleanly than with the copy-paste's from the code base that are currently in there. I'll propose a cleaner example for the story.

Reviewer 4: To contextualize the snippets in @erichgess's post, here is an approximation of the parallelization strategy in our science codes.

```rust
use std::sync::Arc;
use std::collections::HashMap;
use futures::FutureExt;
use futures::future::join_all;
use tokio::runtime::Runtime;

async fn update(runtime: Runtime, solver: Solver, state: State) -> Result<State, Error> {

    let mut stage1 = HashMap::new();
    let mut stage2 = HashMap::new();
    let mut stage3 = HashMap::new();

    // Stage 1: results can be computed independently on each block. Tokio
    // futures must be 'static, so anything moved into the async block has to
    // be cloned first.
    for (index, block) in &state.patches {
        let block = block.clone();
        let solver = solver.clone();

        let future = async move {
            solver.try_stage1(block)
        };
        stage1.insert(index, runtime.spawn(future).map(|f| f.unwrap()).shared());
    }
    let stage1 = Arc::new(stage1);

    // Stage 2: each result requires multiple blocks froms stage 1.
    for (index, _) in &state.patches {
        let solver = solver.clone();
        let stage1 = stage1.clone();

        let future = async move {
            let neighbors = join_all(solver.neighbors_of(index, stage1)).await;
            solver.try_stage2(index, neighbors)
        };
        stage2.insert(index, runtime.spawn(future).map(|f| f.unwrap()).shared());
    }
    let stage2 = Arc::new(stage2);

    // Stage 3: similar to stage 2
    for (index, _) in &state.patches {

    }
    let stage3 = Arc::new(stage3);

    // Final stage: convert HashMap<Future<Result<T, E>>> -> Future<Result<Vec<T>, E>>
    let new_patches: Result<Vec<_>, _> = join_all(stage3.into_values())
        .await
        .into_iter()
        .collect();

    State { patches: new_patches? }
}

```

Once her team began using the new `async` design for their simulations, they noticed an important issue that impacted productivity: compilation time had now increased to between 30 and 60 seconds. The nature of their work requires frequent changes to code and recompilation and 30-60 seconds is long enough to have a noticeable impact on their quality of life.  What her and her team want is for compilation to be 2 to 3 seconds. Barbara believes that the use of `async` is a major contributor to the long compilation times.

This new solution works, but Barbara is not satisfied with how complex her code wound up becoming with the move to `async` and the fact that compilation time is now 30-60 seconds.  The state sharing adding a large amount of cruft with `Arc` and `async` is not well suited for using a dependency graph to schedule tasks so implementing this solution created a key component of her program that was difficult to understand and pervasive. Ultimately, her conclusion was that `async` is not appropriate for parallelizing computational tasks. She will be trying a new design based upon Rayon in the next version of her application.

## 🤔 Frequently Asked Questions

> ### **What are the morals of the story?**

Reviewer 3: I'd like to explore the choice of `tokio` a bit more. I wonder if the choice of that particular runtime led to some of the issues (though it's clear that not all of them came from this choice). Perhaps the lack of a runtime designed for compute bound workloads is another issue here.

Author: That's a very good point: I'll follow up with the source to get more details on how `tokio` was chosen. Personally, I believe that this is either a missing piece of "guidance" in terms of helping people understand the type of concurrency problem they are solving (IO bound, compute bound, etc) and how to get started on the right foot for solving that type of problem.

Author: I don't yet know enough much about this area of the Rust ecosystem. What's the runtime refer to with regards to `tokio`?  The library as a whole, it's API, or the code under the hood that manages threads, task execution, scheduling, etc?

One big problem, from my perspective, is that the semantics of `async` are simply not able to express a complex network of thousands of compute tasks in a simple way.  Trying to write a task until 3 other tasks have completed isn't easy with `async`. In this story, Barbara had to jury rig a dependency-based scheduler into `async` that would block tasks from starting their computations until their dependencies were completed. Put another way, the semantics of `async` are simply meant for  very different type of concurrency problem than the one Barbara was solving.

@jzrake let me know if this is an accurate interpretation of your experience.

Reviewer 3: >  this is either a missing piece of "guidance"

I 100% agree here. The async ecosystem is highly focused on server side scenarios (for the obvious reason that it's a very popular use case), but we need to do better advising users who have different needs. 

> simply meant for very different type of concurrency problem than the one Barbara was solving.

I'm not sure whether I agree or disagree with you. async/await lets you declare dependencies of a certain piece of computation in the same way you would in synchronous code:

```rust
let dep1 = dep1.await;
let dep2 = dep2.await
// now compute based on dep1 and dep2
let answer = compute(dep1, dep2);
```

If you need dependencies to run in parallel you can join them.

```rust
let deps = join_all(dep1, dep2).await;
let answer = compute(deps);
```

Obviously, the experience represents here shows that async/await as it is today does not lead to simple to understand code for compute bound workloads, but I'm just not sure if this is a fundamental property of the computation model or just lack of good guidance and good tooling for this use case.

Author: > I'm not sure whether I agree or disagree with you. async/await lets you declare dependencies of a certain piece of computation in the same way you would in synchronous code:
> If you need dependencies to run in parallel you can join them.
Right, but I'm wondering if that API just doesn't work well when you have thousands of dynamically created tasks that are all joined together in a complex dependency graph. Which makes me wonder if the `async` API assumes that the structure of the code defines the dependencies of the tasks.  For example, if you have tasks X, Y, Z that depend on A, B then you write:

```rust
// define tasks a and b
let deps = join_all(a, b).await;

// define tasks x, y, z
join_all(x, y, z).await;
```
and that might create some friction when trying to write code that coordinates a lot compute based tasks.

> Obviously, the experience represents here shows that async/await as it is today does not lead to simple to understand code for compute bound workloads, but I'm just not sure if this is a fundamental property of the computation model or just lack of good guidance and good tooling for this use case.
Totally agree. One of my take-aways from this story is that it's worth taking some time to see if there is a good solution to Barbara's problem using `async`, if so then we could use that to improve the documentation and guidance, and if not then ask if this is a feature that `async` is missing, or if `async` is not the right tool, etc.

Reviewer: I don't believe tokio led to any issues here, even though i agree that there a futures library designed for computational workflows might help (I would think it wouldn't require deep integration into the executor).

Reviewer: what I mean is: the difficulties I imagine tokio brings is perhaps that its fair scheduler could be more efficient for computation bound workflows, for example, where fairness is not a concern. But that doesn't play into this story, which is more about the challenge of picking the right thing and general "hardness" of async Rust overall. 

I do think a library that had functions or abstractions for doing computational DAGs like this might help, of course, but I don't see why that would be tied closely to the executor (at least, not yet).

Reviewer 4: I think there are barriers to the clarity

1. there are 1000's of tasks, and linking them creates awkward code: not Tokio's fault
2. running the tasks on workers requires `spawn`: also not Tokio's fault (I think)

(1) I believe is just inherent to `async` as a construct. (2) is a problem because it couples the computation to the executor (to perform well, Tokio wants to be the spawner and the executor). If the `async` block had syntax to hint the executor it wants to run on a worker thread, such as

```rust
let y = async cpu_bound {
    calculate(x.await)
};
```

that would alleviate (2). But (1) is still an issue.

Author: To circle back, are there any changes to the story that I should make based on this discussion?

Author: After thinking about it over the weekend, I think this indicated that how I described the chain of events that led using `tokio` was not explicit enough.  So, I've update the story to make that more explicit. The first paragraph of the `Solution Path` section now has a very simple step-by-step for how the character wound up using `tokio` for compute bound parallelization.

Reviewer: I'm quite happy with it.

- `async` looks to be the wrong choice for parallelizing compute bound/computational work
- There is a lack of guidance to help people solving such problems get started on the right foot
- Quality of Life issues (compilation time, type inference on `Result`) can create a drag on users ability to focus on their domain problem

### **What are the sources for this story?**
This story is based on the experience of building the [kilonova](https://github.com/clemson-cal/app-kilonova) hydrodynamics simulation solver.

> ### **Why did you choose *NAME* to tell this story?**

Reviewer 2: Missing the name of the character

Author: Fixed.

I chose Barbara as the primary character in this story because this work was driven by someone with experience in Rust specifically but does not have much systems level experience. Grace was chosen as a supporting character because of that persons experience with C/C++ programming and to avoid repeating characters.

### **How would this story have played out differently for the other characters?**

> For Alan, there's a good chance he would have already had experience working with either async workflows in another language or doing parallelization of compute bound tasks; and so would already know from experience that `async` was not the right place to start.  Grace, likewise, might already have experience with problems like this and would know what to look for when searching for tools. For Niklaus, the experience would probably be the same, as it's very easy to assume that `tokio` is the starting place for concurrency in Rust.

Reviewer 2: I think using a bullet list for the different characters is better readable and lines up with the current style

Author: Fixed.

# 😱 Status quo stories: Barbara Builds a Hydrodynamics Simulator

## 🚧 Warning: Draft status 🚧

This is a draft "status quo" story submitted as part of the brainstorming period. It is derived from real-life experiences of actual Rust users and is meant to reflect some of the challenges that Async Rust programmers face today. 

If you would like to expand on this story, or adjust the answers to the FAQ, feel free to open a PR making edits (but keep in mind that, as they reflect peoples' experiences, status quo stories [cannot be wrong], only inaccurate). Alternatively, you may wish to [add your own status quo story][htvsq]!

## The story
### Problem

> Barbara needed to build a tool to solve hydrodynamics simulations; there is a common method for this that subdivides a region into a grid and computes the solution for each grid patch. All the patches in a grid for a point in time are independent and can be computed in parallel, but they are dependent on neighboring patches in the previously computed frame in time.  This is a well known computational model and the patterns for basic parallelization are well established.

Reviewer: ```suggestion
Barbara is a professor of physics at the University of Rustville. She needed to build a tool to solve hydrodynamics simulations; there is a common method for this that subdivides a region into a grid and computes the solution for each grid patch. All the patches in a grid for a point in time are independent and can be computed in parallel, but they are dependent on neighboring patches in the previously computed frame in time.  This is a well known computational model and the patterns for basic parallelization are well established.
```

Author: Done.

> Barabara wanted to write a performant tool to compute the solutions to the simulations of her research.  She chose Rust because she was already familiar with it and it had good qualities for writing performant code. After implementing the core mathematical formulas, Barbara began implementing the parallelization architecture. 

Reviewer: ```suggestion
Barabara wanted to write a performant tool to compute the solutions to the simulations of her research.  She chose Rust because she needed high performance but she also wanted something that could be maintained by her students, who are not professional programmers. Rust's safety guarantees give her confidence that her results are going to be corrupted from data races or other programming errors. After implementing the core mathematical formulas, Barbara began implementing the parallelization architecture. 
```

Author: Done
